### ISSUE TITLE
PrivateBin NextCloud App

### ISSUE THREAD

**Issue by JPustkuchen** [PrivateBin/PrivateBin#967]
Hi and thank you very, very much for this great app! This is a feature request to provide a NextCloud App for PrivateBin, which might push the use of PrivateBin a lot and make it easier for non-technical users to use its great technology.

NextCloud is a self-hosted cloud platform for privacy and security in the focus.

There's already a request for that at NextCloud:
https://help.nextcloud.com/t/pastebin-app-for-nextcloud/124233

**Comment by rugk** [comment]
The use case here would be to simplify hosting, would it? So just the usual PB software as an app and maybe have a 

Or how else would you imagine a Nextcloud integration?

**Comment by JPustkuchen** [comment]
Exactly @rugk !

**Comment by elrido** [comment]
I guess that would require a separate git project, since I assume turning this into a nextcloud app would require integrating into Nexclouds routing / app mechanism instead of running stand-alone with our own routing and UI and having a storage backend that stores data inside nextcloud using it's APIs.

So I'd consider this to be out of scope for this project, but if you (or anyone else) created such a fork, we can host it under the umbrella of the [privatebin organization](https://github.com/PrivateBin/).

**Comment by damienvancouver** [comment]
@elrido you're correct in your assessment that it should be a separate thing from the project, but rather than a fork it should be a related project that bundles the upstream.

TL;DRs:  

1. the use case is to extend PB as a simple one click install to the 100's of millions of Nextcloud installs out there, and their users.

2. Privatebin being pure PHP is an exceptional candidate for easy nextcloud integration, it would be an easy project to accomplish (comparitively) and it would just ship the upstream PB code, not fork it.  Details on how that would look.


1 First off, the use case / why this would be an important step:  Nextcloud originally started out as a Dropbox clone (with big emphasis on desktop clients syncing files).  But times have changed and it has massively grown in functionality, to the point where it works as a browser based home screen / dashboard for users instead of a small single tool.  The Nextcloud home screen/dashboard has everything a user needs to get work done, with a toolbar of apps they can run across the top.  This is intended as a replacement for Google Workspace / Office 365 etc, so apps include things like email client / calendar / tasks / other groupware, a text/video Talk client, rich realtime collab markdown file editing, rich realtime collab office document editing (with collabora office or onlyoffice integration) plus integration with hundreds of other open source apps.    With all those in one place, users are inclined to never leave that homescreen environment.  So to make more open source projects available in that world, the community has set to work integrating as many as possible.  

So the use case would be"
- to make Privatebin available as a simple "app install" inside Nextcloud for administrators.  They can "install" it by going to the Nextcloud App store and hitting install and that's it.
- to perhaps have PB become the de-facto private note sharing app inside Nextcloud
- if that happens then you've extended your user base by potentially hundreds of millions, enhancing their security and empowering them to host their own solution.  Which is pretty nifty!   

2. Secondly, this one begs to be done because it is so easy.  Nextcloud is (mostly) pure PHP.  The simplest apps to integrate are also pure PHP as they require no additional setup (there are no daemons must run or have supervised starting and no webserver reverse proxy config must be added).   Many projects have to deal with both those things, but since Privatebin is PHP it can easily be installed on any Nextcloud without any sort of server or hosting admin privileges required.  A Nextcloud admin just installs it from the app store and uses it.  

In most cases Nextcloud integratoin means simply making a wrapper app for the upstream code, which does 2 main things. 
Thing #1: It makes an app icon for the nextcloud toolbar so there is something for users to click to get it open. (It can also assocaite with file extensions and open that way).   
Thing #2: it presents the UX of the app inside the Nextcloud environment (the toolbar is up top and the app screen is at the bottom, basically).   That's about it for a lot of cases, though some people start re-implementing UX in their nextcloud apps where it makes sense and where the app supports that architecture.  Most though are simple things that provide icon and wrap the UX.

So logistically what a Nextcloud app of privatebin could probably look like would be the Nextcloud app PHP files and classes with an upstream vendor folder that just included the relevant Privatebin files to present them in the nextcloud UX.  With luck little or no modification of upstream PB would be required to accomplish this.  That is where coopeation between the projects would be key, there may be some (hopefully minor and non impacting) refactoring of the PB codebase needed to make it so upstream can run unforked (beneficial for so many reasons of course).  

The NC app does need someone babysitting it though and compatibility testing, plus releasing new versions or it will show as incompatible on the nextcloud app store.   IMO the core upstream devs should not be getting involved in maintaining the integration app as well, they should be focused on doing what they do best and someone invested in the Nextcloud community should be babysitting the integration.     I am not that someone yet but I may one day be, when I have time to get into NC app development.  I could definitely use this as an integrated app in the offering I make available to customers (who I'm trying to wean off corporate and ad-riddled "free" services).



I am not (yet) a Nextcloud app developer but I plan to get into it sometime.  If no one else has taken up the charge by then I'd be willing to look at it.  I agree that

**Comment by tunloop** [comment]
I recently added privatebin to my homelab and I actually got a sort-of integration into nextcloud. 
I wanted to share this because it seems like an easier way of full integration (albeit with more manual config than an app).

I threw down the privatebin web files into its own directory next to the nextcloud folder. My reasoning was this application is pretty simple/lightweight so having it just sit on the same backend server as nextcloud is okay in my book.
So something like:
/var/www
  /nextcloud
  /privatebin

Then added a location block to my reverse proxy under the nextcloud server block
        # PrivateBin sub url site
        location /privatebin/ {
                proxy_pass https://192.168.20.5/; #IP to nextcloud backend server
                proxy_set_header X-Forwarded-For $remote_addr;
                proxy_set_header Host "privatebin.local.net";
                access_log off; #optional
        }

And its own site config on the backend nginx server:
server {
    listen 443 ssl;
    http2 on;
    include snippets/self-signed.conf;
    include snippets/ssl-params.conf;
    server_name privatebin.local.net; #filter requests from reverse proxy by host header

    # set max upload size
    client_max_body_size 512M;
    fastcgi_buffers 64 4K;

    root /var/www/privatebin;
    index index.php /index.php$request_uri;

    location ~ \.(?:css|js|svg|gif|png|jpg|ico)$ {
        try_files $uri /index.php$request_uri;
        expires 6M;         # Cache-Control policy
        access_log off;     # Optional: Don't log access to assets
    }

    location ~ \.php(?:$|/) {
        fastcgi_split_path_info ^(.+?\.php)(/.*)$;
        set $path_info $fastcgi_path_info;

        try_files $fastcgi_script_name =404;

        include fastcgi_params;
        fastcgi_param SCRIPT_FILENAME $document_root$fastcgi_script_name;
        fastcgi_param PATH_INFO $path_info;
        fastcgi_param HTTPS on;

        fastcgi_pass php-handler; #php .sock global var

        fastcgi_intercept_errors on;
        fastcgi_request_buffering off;
    }
}

Then I used the nextcloud app "External Sites" by Joas Schilling to add the link https://nextcloud.domain.com/privatebin/ and allow embedding.
Then I went in to the cfg/conf.php and removed the frame-ancestors part of the csp header. (Embedding didn't work when frame-ancestors was set to the domain/url of nextcloud unfortunately)
Then in lib/Controller.php I set x-frame-options to sameorigin. (For me I actually commented out most of these options, as they are already set by my reverse proxy - would be nice to have an option in conf.php to change these : ) )

Now I have a section in the top bar of nextcloud that I can click and it loads the privatebin interface inside the nextcloud interface.
Another idea here is that this could be improved by adding a first party theme template to privatebin to make it blend in with nextcloud themes/colors/layout. 

I suppose this could still be packaged with a nextcloud app to make the backend configurations less manual. But I hope this little blurb can help people implement a partial-jank solution in the mean time.

Let me know what you think!
